### ISSUE TITLE
recording more than 4 event counters

### ISSUE THREAD

**Issue by anwesha94** [jcsaezal/pmctrack#6]
Hi, I am trying to get stats for multiple counters. I am getting the following in the output file:
```
[Event-to-counter mappings]

pmc1=instr(0),llc_misses(1)

pmc2=cycles(0),bus_accesses(1)

pmc3=l2_cache_accesses(0)

pmc4=l1_data_accesses(0)

[Event counts]

nsample    cpu coretype expid      event          pmc1          pmc2          pmc3          pmc4

      1      0        0     0       tick         16201        122948          2471          5695 

      2      1        0     0       tick        897005       4291915         11347        206329 

      3      2        0     0       tick         10178         83281          1693          4096 

      4      3        0     0       tick       1626577       5142808         69491        685774 

      5      4        1     0       tick         17712         67290          2093          8100 

      6      5        1     0       tick           115          1929            78            44 

      7      6        1     0       tick         46568        167465          4011         17111 

      8      7        1     0       tick             0             0             0             0 

      9      0        0     1       tick             0             0             -             - 

     10      1        0     1       tick          3331         82381             -             - 

     11      2        0     1       tick             0             0             -             - 

     12      3        0     1       tick          4179         18810             -             - 

     13      4        1     1       tick           106           493             -             - 

     14      5        1     1       tick             0             0             -             - 

     15      6        1     1       tick           486          3060             -             - 

     16      7        1     1       tick             0             0             -             - 

     17      0        0     0       tick             0             0             0             0 

     18      1        0     0       tick       2372160      10162903         50591        540864 

     19      2        0     0       tick             0             0             0             0 

     20      3        0     0       tick        646372       1332981         15585        228872 

     21      4        1     0       tick         50142         69657          1440         20128 

     22      5        1     0       tick             0             0             0             0 

     23      6        1     0       tick        189017        381921         14571         55095 

     24      7        1     0       tick             0             0             0             0 

     25      0        0     1       tick             0             0             -             - 

     26      1        0     1       tick          1239        499002             -             - 

     27      2        0     1       tick             0             0             -             - 

     28      3        0     1       tick             0             0             -             - 

     29      4        1     1       tick            45           208             -             - 

     30      5        1     1       tick             0             0             -             - 

     31      6        1     1       tick           757          4199             -             - 

     32      7        1     1       tick             0             0             -             - 
```

Does this mean that I can not record more than 4 counters at one particular instant?
Or expid 0 and 1 both are recorded at the same time but displayed separately? what does expid represent?
Can you please explain how to  interpret this output?
Thank you.

**Comment by sirfonzie** [comment]
nsample    cpu coretype expid      event          pmc1          pmc2          pmc3          pmc4
            1      0        0               0       tick            16201        122948          2471          5695 

The above means pmc1=instr, pmc2=cycles, pmc3=l2_cache_accesses and pmc4=l1_data_accesses.

nsample    cpu coretype expid      event          pmc1          pmc2          pmc3          pmc4
            9      0        0                1       tick                   0             0             -             - 

The above means, pmc1=llc_misses, pmc2=bus_accesses while pmc3 and pmc4 not used.

Both samples above are for cpu 0, core 0.
At least thats my understanding.

**Comment by anwesha94** [comment]
I understand that. My question is are they at the same instant of time?

**Comment by jcsaezal** [comment]
@anwesha94 

Little cores in the Odroid XU3 (or XU4) board are equipped with 4 general-purpose  (GP) counters, hence the 4-counter limit. Big cores, by contrast, integrate 6 GP counters. To verify the number of counters of each Performance Monitoring Unit (PMU) on a platform, you can check the contents of the `/proc/pmc/info` file (`nr_gp_pmcs` property):

```bash
$ cat /proc/pmc/info
*** PMU Info ***
nr_core_types=2
[PMU coretype0]
pmu_model=armv7.cortex_a7
nr_gp_pmcs=4
nr_ff_pmcs=1
pmc_bitwidth=32
flags=[pmc(8),usr,os,ebs(32),coretype/g]
[PMU coretype1]
pmu_model=armv7.cortex_a15
nr_gp_pmcs=6
nr_ff_pmcs=1
pmc_bitwidth=32
flags=[pmc(8),usr,os,ebs(32),coretype/g]
***************
*** Monitoring Module ***
counter_used_mask=0x0
nr_experiments=0
nr_virtual_counters=0
```
As stated in [PMCTrack documentation](https://pmctrack.dacya.ucm.es/getting-started/):

> In case a specific processor model does not integrate enough PMCs to monitor a given set of events at once, the user can turn to PMCTrack’s event-multiplexing feature. This boils down to specifying several event sets by including multiple instances of the -c switch in the command line. In this case, the various events sets will be collected in a round-robin fashion and a new expid field in the output will indicate the event set a particular sample belongs to. 

I guess that the command you might have used is as follows:

```
$ pmctrack -c instr,cycles,l2_cache_accesses,l1_data_accesses,llc_misses,bus_accesses ....
```
In that case PMCTrack detects that you are trying to use more than 4 counters (the most restrictive scenario in little cores), and breaks down the event set automatically into two events sets with hardware events that can be monitored simultaneously. Note that each event set is assigned a different expid value (number in parentheses shown in the "Event-to-counter mappings" section). So that is equivalent to invoking `pmctrack` as follows: 

```   
$ pmctrack -c instr,cycles,l2_cache_accesses,l1_data_accesses -c llc_misses,bus_accesses ....
```
In summary, that just is like specifying two separate sets of events to be monitored one after another in a round-robin fashion. 

--Juan Carlos
